### ISSUE TITLE
Add guide on working with Effection and TypeScript

### ISSUE THREAD

**Issue by cowboyd** [thefrontside/effection#579]
One of the gaps in Effection is that it has some idiosyncrasies in working with TypeScript. In particular with the way TS handles generators, there is no way for the right hand side of a `yield` expression to influence the inferred type of the left hand side. As a result, you need to explicitly annotate your intermediate values, otherwise TS infers them as being of type `any`

```typescript
// `bare` is typed as `any`
let bare = yield Promise.resolve(10);

// to get type support you need to explicitly annotate
let num: number = yield Promise.resolve(10);
```

This can be problematic since you can actually get it wrong and get the compiler out of sync with the actual code which can result in mystifying runtime errors.

```typescript
// this will type-check just fine
let num: number = yield Promise.resolve('not a number!');
```

The fact that there is not inference is confusing [as seen here](https://github.com/thefrontside/actions/pull/65#discussion_r727732663) and so we should arm folks with the knowledge a) not to be suprised by it and b) how to work with it, because it's something that will likely happen very soon in their journey.

**Comment by jorgelainfiesta** [comment]
I see. What would be required for TS to be able to infer the type correctly from the yield statement?

**Comment by cowboyd** [comment]
It would require a new release of TypeScript, but given the response from one of the folks on the core team, they are still not sure how to approach it yet https://github.com/microsoft/TypeScript/issues/38534#issuecomment-650467285

**Comment by cowboyd** [comment]
See also https://github.com/microsoft/TypeScript/issues/36967

I also have a prototype of Effection that works with TypeScript out of the box: https://github.com/cowboyd/nano but it is only a very recent discovery, and would imply some pretty major breaking changes, and so we reasonably wouldn't be able to leverage it for at least a year.

**Comment by jorgelainfiesta** [comment]
Gotcha thanks, I was asking to figure out the way of explaining the limitation in the guides.

**Comment by cowboyd** [comment]
Closed by #599
